src/storage.rs

```diff
@@ -0,0 +1,151 @@
+//! Storage
+/// Traits to allow on and off board storage deivces to read and write data
+/// Allows for Read Only Memory as well as Read and Write Memory
+use nb;
+
+/// Address represents an unsigned integer. This allows for devices that have bigger or smaller address spaces than the host.
+pub struct Address<U>(pub U);
+/// Address Offset represents an unsigned integer that is used as an optional offset from the base address.
+pub struct AddressOffset<U>(pub U);
+
+use core::ops::{Add, Sub};
+
+/// Implement add for the Address and AddressOffset Types.
+impl<'a, 'b, U> Add<&'b AddressOffset<U>> for &'a Address<U>
+where
+    U: Add<U, Output = U> + Copy,
+{
+    type Output = Address<U>;
+
+    fn add(self, other: &'b AddressOffset<U>) -> Address<U> {
+        Address(self.0 + other.0)
+    }
+}
+
+/// Implement subtraction for the Address and AddressOffset Types.
+impl<'a, 'b, U> Sub<&'b AddressOffset<U>> for &'a Address<U>
+where
+    U: Sub<U, Output = U> + Copy,
+{
+    type Output = Address<U>;
+
+    fn sub(self, other: &'b AddressOffset<U>) -> Address<U> {
+        Address(self.0 - other.0)
+    }
+}
+
+/// Page represents an unsigned integer that is a Page ID in the device memory space.
+pub struct Page<U>(pub U);
```

Comment:
What do you mean by "`Page`"? For NOR flashes this is typically the maximum number of bytes you can write in one go. Since you used it further down in `try_erase_page` I suspect you mean the minimum erase size of the device. For NOR flashes this is typically a block. I thought about having something like that too, but erase sizes in flashes can be quite complicated, so I didn't know how to model that.